### PR TITLE
Set ClusterDataReady false if PV exists, is bound, and claim's deletion timestamp is non-zero

### LINKED ISSUE #943: PVC remain in pending state after successful failover
Steps to reproduce:
Initiate failover of an application on cluster where earlier PVCs in terminating state which is needed while changing ReplicationState of vrg to secondary.
After successful failover, PVCs in pending state.
Observed that UID field of PV in 'ClaimRef' not refreshed.

### PR DESCRIPTION
**Pull request by hatfieldbrian**
Fixes #943

### CODE REVIEW

**Review by raghavendra-talur — APPROVED** [review]

**Comment by hatfieldbrian** [comment]
Still working on a unit test, but pushed the controller patch in the meantime for early review

File: controllers/vrg_volrep.go [hunk 1]
```diff
@@ -1953,41 +1959,46 @@ func (v *VRGInstance) updateExistingPVForSync(pv *corev1.PersistentVolume) error
...
+	if existingPV.Status.Phase == corev1.VolumeBound {
+		var pvc corev1.PersistentVolumeClaim
 
-		return nil
-	}
+		pvcNamespacedName := types.NamespacedName{Name: pv.Spec.ClaimRef.Name, Namespace: pv.Spec.ClaimRef.Namespace}
+
+		err := v.reconciler.Get(v.ctx, pvcNamespacedName, &pvc)
+		if err != nil {
+			return fmt.Errorf("failed to get PV %s claim %s: %w", existingPV.GetName(), pvcNamespacedName.String(), err)
```
Comment (ShyamsundarR) [inline controllers/vrg_volrep.go:1978]: (nits) Thinking if the error message can be changed like so:

"found bound PV %s to claim %s but unable to validate claim exists: %w"

and something similar in 1982 stating deleted claim (maybe just adding "bound" word between "existing bound PV..."

Reason being, we expect the situation to be rectified as the PV should reach the Released state at some point in time. So failing to get a claim that does not exist is fine. (and that this message may/will reach the VRG conditions)

**Review by ShyamsundarR — COMMENTED** [review]
LGTM! Checking other metro cases that had fixes around this area before merge etc.

File: controllers/vrg_volrep.go (hunk 1 above)
Comment (hatfieldbrian) [reply controllers/vrg_volrep.go]: Fixed

**Review by ShyamsundarR — APPROVED** [review]
Will wait for one more ack from @raghavendra-talur based on the refreshed code (and maybe a metro test run), before a merge.

**Review by raghavendra-talur — APPROVED** [review]

**Comment by raghavendra-talur** [comment]
LGTM, I have reviewed the code. I haven't run a complete metro test yet but we can merge the PR.

File: controllers/vrg_volrep.go [hunk 2]
```diff
@@ -1899,16 +1890,31 @@ func (v *VRGInstance) checkPVClusterData(pvList []corev1.PersistentVolume) error
...
+	validateExistingObject func(*ObjectType) error,
+) (int, error) {
 	numRestored := 0
 
-	for _, obj := range objList {
-		v.cleanupForRestore(obj)
-		v.addRestoreAnnotation(obj)
+	for i := range objList {
+		object := &objList[i]
+		objectCopy := &*object
```
Comment (BenamarMk) [inline controllers/vrg_volrep.go:1909]: Why is this a copy? I think objectCopy == object == &objList[i]. They all point to the same memory location.

File: controllers/vrg_volrep.go [hunk 3]
```diff
@@ -1899,16 +1890,31 @@ func (v *VRGInstance) checkPVClusterData(pvList []corev1.PersistentVolume) error
 	return nil
 }
 
-func (v *VRGInstance) restoreClusterDataObjects(objList []client.Object, objType string) (int, error) {
+func restoreClusterDataObjects[
+	ObjectType any,
```
Comment (BenamarMk) [inline controllers/vrg_volrep.go:1894]: I have never seen this block of code before with a function.
I wonder what `func f[x,y]()` mean? I probably should just google it, but I can't even think of what to ask.

Comment (hatfieldbrian) [reply controllers/vrg_volrep.go:1894]: It's a [generic](https://go.dev/doc/tutorial/generics), similar to a C++ template.  A function is written once but supports multiple types.  It resolves types at compile-time which is preferable to runtime type checking.

**Pull request merged**